File: round14c-delivery/py/openai/src/agentloop_openai/_ask.py
"""
Core helper that wraps a single OpenAI chat-completions call in the
AgentLoop cycle: search → inject → call → log_turn.

Exported as `ask_with_agentloop` for callers who want explicit control.
The `wrap_openai` Proxy in wrap.py builds on top of this.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import TYPE_CHECKING, Any, Callable

from agentloop import AgentLoop, Memory

if TYPE_CHECKING:
    from openai import OpenAI


# ---------------------------------------------------------------------------
# Configuration dataclasses
# ---------------------------------------------------------------------------


@dataclass
class PerCallOptions:
    """Per-call overrides passed via the `agentloop` keyword to wrapped calls.

    Every field is optional. A wrapped call with no `agentloop` kwarg at
    all still works — the wrap-time config applies.
    """

    user_id: str | None = None
    session_id: str | None = None
    signals: dict[str, Any] | None = None
    metadata: dict[str, Any] | None = None
    #: Skip AgentLoop entirely — no search, no log_turn. Useful for
    #: health checks, system calls, or any turn you don't want reviewed.
    skip: bool = False
    #: Skip only memory retrieval for this call. Log_turn still fires.
    #: Pass False to disable, or a dict like {"limit": 5, "tags": [...]}
    #: to override defaults.
    search: bool | dict[str, Any] = True


@dataclass
class WrapOptions:
    """Configuration passed to `wrap_openai(client, loop=..., ...)`."""

    loop: AgentLoop

    #: How to inject retrieved memories into the messages array. Default
    #: prepends a "Trusted facts from past corrections:" block to the
    #: first system message (or creates one if absent).
    inject_memories: Callable[[list[Memory], list[dict[str, Any]]], list[dict[str, Any]]] | None = None

    #: Auto-detect signals from the assistant's response. Merged with
    #: any per-call signals the caller passed.
    detect_signals: Callable[[str, str, list[Memory]], dict[str, Any] | None] | None = None

    search_limit: int = 3
    search_tags: list[str] = field(default_factory=list)

    #: If True, log_turn is skipped when no signals fired. Default False
    #: — log every turn so reviewers see baseline behavior.
    only_log_when_signaled: bool = False


# ---------------------------------------------------------------------------
# Default memory injector
# ---------------------------------------------------------------------------


def default_inject_memories(
    memories: list[Memory], messages: list[dict[str, Any]]
) -> list[dict[str, Any]]:
    """Default: append a "Trusted facts" block to the first system message."""
    if not memories:
        return messages

    facts_block = "\n\nTrusted facts from past corrections:\n" + "\n".join(
        f"- {m.fact}" for m in memories
    )

    for i, msg in enumerate(messages):
        if msg.get("role") == "system":
            updated = list(messages)
            existing = msg.get("content", "") or ""
            if isinstance(existing, str):
                updated[i] = {**msg, "content": existing + facts_block}
            return updated

    # No system message — prepend one.
    return [
        {"role": "system", "content": "You are a helpful assistant." + facts_block},
        *messages,
    ]


# ---------------------------------------------------------------------------
# Extract the question from an OpenAI messages array
# ---------------------------------------------------------------------------


def extract_question(messages: list[dict[str, Any]]) -> str:
    """Last user message, text content only."""
    for msg in reversed(messages):
        if msg.get("role") != "user":
            continue
        content = msg.get("content")
        if isinstance(content, str):
            return content
        if isinstance(content, list):
            # Multi-part: concat text parts, skip image parts
            parts = [
                p["text"]
                for p in content
                if isinstance(p, dict) and p.get("type") == "text"
            ]
            return "\n".join(parts)
    return ""


def extract_answer(completion: Any) -> str:
    """Pull text content from a ChatCompletion response object.

    Accesses `.choices[0].message.content` which is the OpenAI SDK's
    object shape. Tolerates missing fields by returning empty string.
    """
    try:
        choices = getattr(completion, "choices", None) or completion.get("choices", [])
        if not choices:
            return ""
        msg = getattr(choices[0], "message", None) or choices[0].get("message", {})
        content = getattr(msg, "content", None) if hasattr(msg, "content") else msg.get("content")
        return content or ""
    except (AttributeError, KeyError, IndexError, TypeError):
        return ""


# ---------------------------------------------------------------------------
# Main entry point — non-streaming
# ---------------------------------------------------------------------------


def ask_with_agentloop(
    client: OpenAI,
    *,
    messages: list[dict[str, Any]],
    per_call: PerCallOptions | None = None,
    config: WrapOptions,
    **create_kwargs: Any,
) -> Any:
    """Run one OpenAI chat-completions call through the full AgentLoop cycle.

    Use this directly if you want explicit control (custom signal
    detection, conditional skipping, non-chat endpoints). The
    `wrap_openai` Proxy builds on top of this for the drop-in case.
    """
    per_call = per_call or PerCallOptions()

    if per_call.skip:
        return client.chat.completions.create(messages=messages, **create_kwargs)

    question = extract_question(messages)

    # ---- 1. Retrieve memories ----
    memories: list[Memory] = []
    if per_call.search is not False and question:
        search_config = per_call.search if isinstance(per_call.search, dict) else {}
        memories = config.loop.search(
            question,
            limit=search_config.get("limit", config.search_limit),
            user_id=per_call.user_id,
            tags=search_config.get("tags") or (config.search_tags or None),
        )

    # ---- 2. Inject memories into prompt ----
    injector = config.inject_memories or default_inject_memories
    augmented_messages = injector(memories, messages)

    # ---- 3. Call the LLM ----
    completion = client.chat.completions.create(
        messages=augmented_messages, **create_kwargs
    )
    answer = extract_answer(completion)

    # ---- 4. Detect signals + log_turn ----
    auto_signals = (
        config.detect_signals(question, answer, memories)
        if config.detect_signals
        else None
    ) or {}
    merged_signals: dict[str, Any] = {**auto_signals, **(per_call.signals or {})}

    should_log = not config.only_log_when_signaled or len(merged_signals) > 0

    if should_log and question and answer:
        config.loop.log_turn(
            question=question,
            agent_response=answer,
            user_id=per_call.user_id,
            session_id=per_call.session_id,
            signals=merged_signals or None,
            metadata=per_call.metadata,
        )

    return completion

File: round14c-delivery/js/openai/src/ask.ts
/**
 * Core helper that wraps a single OpenAI chat-completions call in the
 * full AgentLoop cycle: search → inject → call → logTurn.
 *
 * This is the low-level API the Proxy wrapper builds on. Exported
 * directly for callers who want explicit control — custom signal
 * detection, non-chat endpoints, conditional skipping, etc.
 *
 * Keeps no state: each call is independent. Same function handles
 * multi-turn chats (caller passes the full message history) — we only
 * retrieve against the last user message, which matches how chatbots
 * actually get used.
 */

import type { AgentLoop, Memory, TurnSignals } from "@agentloop-sdk/core";
import type {
  ChatCompletionCreateParamsNonStreaming,
  ChatCompletion,
  ChatCompletionMessageParam,
} from "openai/resources/chat/completions";
import type OpenAI from "openai";

/**
 * Extra options a caller can pass per-call alongside their normal OpenAI
 * chat-completions params. Everything here is optional; a call with no
 * `agentloop` field at all still works (config from `wrapOpenAI` applies).
 */
export interface PerCallOptions {
  /** Scope memory search to this end-user (matches signals.user_id). */
  userId?: string;
  /** Pass through to logTurn() unchanged. */
  sessionId?: string;
  /**
   * Signals to attach to the logged turn. Merged on top of any signals
   * auto-detected by the wrapper (see `detectSignals` option).
   */
  signals?: TurnSignals;
  /** Arbitrary metadata stored with the turn. */
  metadata?: Record<string, unknown>;
  /**
   * Skip the AgentLoop pre/post hooks for this specific call. Useful for
   * system messages, health checks, or any call you don't want reviewed.
   * The underlying OpenAI call still runs normally.
   */
  skip?: boolean;
  /**
   * Override the memory search behavior for this call.
   * - `undefined` (default): use the default (search on, inject on, up to 3)
   * - `false`: don't search for this call
   * - a shape: override the defaults
   */
  search?: false | { limit?: number; tags?: string[] };
}

/**
 * Configuration for `askWithAgentLoop` and for `wrapOpenAI`. Same shape
 * because the wrapper stores these and passes them through.
 */
export interface AgentLoopOpenAIOptions {
  /** The AgentLoop client. Required. */
  loop: AgentLoop;
  /**
   * How to inject retrieved memories into the system prompt. Default
   * prepends an "Trusted facts from past corrections:" block to the first
   * system message (or creates one if absent). Override for a different
   * format, different position, different language.
   */
  injectMemories?: (memories: Memory[], messages: ChatCompletionMessageParam[]) => ChatCompletionMessageParam[];
  /**
   * Auto-detect signals from the assistant's response. Runs after the
   * LLM call, before logTurn. Return `undefined` to skip auto-detection.
   * Merged with any per-call signals the caller passed.
   */
  detectSignals?: (question: string, answer: string, memories: Memory[]) => TurnSignals | undefined;
  /**
   * How many memories to retrieve per call by default. Override per-call
   * via `agentloop.search.limit`. Defaults to 3.
   */
  searchLimit?: number;
  /**
   * Tags to apply to every memory search. Override per-call via
   * `agentloop.search.tags`.
   */
  searchTags?: string[];
  /**
   * If true, logTurn() is skipped when no signals fired AND no random
   * sample rate triggered. Default false — we log every turn so reviewers
   * see baseline agent behavior, not just failures. Set true for
   * high-traffic apps where you only want to see flagged turns.
   */
  onlyLogWhenSignaled?: boolean;
}

// ---------------------------------------------------------------------------
// Default injection: prepend a "Trusted facts" block to system prompt.
// ---------------------------------------------------------------------------

function defaultInjectMemories(
  memories: Memory[],
  messages: ChatCompletionMessageParam[]
): ChatCompletionMessageParam[] {
  if (memories.length === 0) return messages;

  const factsBlock =
    "\n\nTrusted facts from past corrections:\n" +
    memories.map((m) => `- ${m.fact}`).join("\n");

  // Find the first system message and append; if none exists, prepend one.
  const firstSystemIdx = messages.findIndex((m) => m.role === "system");
  if (firstSystemIdx === -1) {
    return [
      { role: "system", content: "You are a helpful assistant." + factsBlock },
      ...messages,
    ];
  }

  const updated = [...messages];
  const systemMsg = updated[firstSystemIdx];
  if (systemMsg && systemMsg.role === "system") {
    const existingContent =
      typeof systemMsg.content === "string" ? systemMsg.content : "";
    updated[firstSystemIdx] = {
      ...systemMsg,
      content: existingContent + factsBlock,
    };
  }
  return updated;
}

// ---------------------------------------------------------------------------
// Extract the "question" from a chat messages array — the last user msg.
// ---------------------------------------------------------------------------

function extractQuestion(messages: ChatCompletionMessageParam[]): string {
  // Scan backwards for the most recent user message. Tool/assistant/system
  // messages in between are part of the conversation scaffolding, not the
  // question we want to retrieve against.
  for (let i = messages.length - 1; i >= 0; i--) {
    const m = messages[i];
    if (m && m.role === "user") {
      if (typeof m.content === "string") return m.content;
      // Multi-part messages: concatenate text parts. (Image parts skipped
      // — we don't have a story for multimodal memory retrieval yet.)
      if (Array.isArray(m.content)) {
        return m.content
          .filter((p): p is { type: "text"; text: string } => p.type === "text")
          .map((p) => p.text)
          .join("\n");
      }
    }
  }
  return "";
}

// ---------------------------------------------------------------------------
// Extract the answer text from a ChatCompletion response.
// ---------------------------------------------------------------------------

export function extractAnswer(completion: ChatCompletion): string {
  const msg = completion.choices[0]?.message;
  if (!msg) return "";
  if (typeof msg.content === "string") return msg.content;
  return "";
}

// ---------------------------------------------------------------------------
// Main entry point (non-streaming).
// ---------------------------------------------------------------------------

export async function askWithAgentLoop(
  openai: OpenAI,
  params: ChatCompletionCreateParamsNonStreaming,
  perCall: PerCallOptions | undefined,
  config: AgentLoopOpenAIOptions
): Promise<ChatCompletion> {
  if (perCall?.skip) {
    // Full bypass — no search, no logTurn, pure passthrough.
    return openai.chat.completions.create(params);
  }

  const question = extractQuestion(params.messages as ChatCompletionMessageParam[]);

  // ---- 1. Retrieve memories ----
  let memories: Memory[] = [];
  if (perCall?.search !== false && question) {
    const searchLimit =
      (perCall?.search && perCall.search.limit) ?? config.searchLimit ?? 3;
    const searchTags =
      (perCall?.search && perCall.search.tags) ?? config.searchTags;
    memories = await config.loop.search(question, {
      limit: searchLimit,
      ...(perCall?.userId && { user_id: perCall.userId }),
      ...(searchTags && searchTags.length > 0 && { tags: searchTags }),
    });
  }

  // ---- 2. Inject into prompt ----
  const injector = config.injectMemories ?? defaultInjectMemories;
  const augmentedMessages = injector(
    memories,
    params.messages as ChatCompletionMessageParam[]
  );

  // ---- 3. Call the LLM ----
  const completion = await openai.chat.completions.create({
    ...params,
    messages: augmentedMessages,
  });
  const answer = extractAnswer(completion);

  // ---- 4. Detect signals + log ----
  const autoSignals = config.detectSignals
    ? config.detectSignals(question, answer, memories) ?? {}
    : {};
  const mergedSignals: TurnSignals = { ...autoSignals, ...(perCall?.signals ?? {}) };

  const shouldLog =
    !config.onlyLogWhenSignaled || Object.keys(mergedSignals).length > 0;

  if (shouldLog && question && answer) {
    // Fire-and-forget. logTurn already degrades gracefully on error, so
    // awaiting this only delays the return; not awaiting would lose
    // backpressure and make testing harder. We await.
    await config.loop.logTurn(question, answer, {
      ...(perCall?.userId && { user_id: perCall.userId }),
      ...(perCall?.sessionId && { session_id: perCall.sessionId }),
      ...(Object.keys(mergedSignals).length > 0 && { signals: mergedSignals }),
      ...(perCall?.metadata && { metadata: perCall.metadata }),
    });
  }

  return completion;
}
